### PR TITLE
sampler: rewrite RateByServiceSampler without using Lock

### LINKED ISSUE #935: Stuck in "Waiting for GIL"
Hi,

We found a thread that seems blocked forever:

```
gdb python 30107
```

Here's the single thread:

```
(gdb) info threads
  Id   Target Id         Frame
* 1    Thread 0x7f1dd852e4c0 (LWP 30107) "/opt/simon/rele" 0x00007f1dd7d2e146 in do_futex_wait.constprop () from /lib64/libpthread.so.0
```

Here's the backtrace:

```
raceback (most recent call first):
  Waiting for the GIL
  File "/opt/simon/releases/b038662ce7d140609b1e4883a3ea0edf77851537/venv/lib/python2.7/site-packages/ddtrace/sampler.py", line 88, in sample
    with self._lock:
  File "/opt/simon/releases/b038662ce7d140609b1e4883a3ea0edf77851537/venv/lib/python2.7/site-packages/ddtrace/tracer.py", line 251, in start_span
    if self.priority_sampler.sample(span):
  File "/opt/simon/releases/b038662ce7d140609b1e4883a3ea0edf77851537/venv/lib/python2.7/site-packages/ddtrace/tracer.py", line 376, in trace
    span_type=span_type,
  File "/opt/simon/releases/b038662ce7d140609b1e4883a3ea0edf77851537/venv/lib/python2.7/site-packages/ddtrace/contrib/django/cache.py", line 56, in wrapped
    with tracer.trace('django.cache', span_type=TYPE, service=cache_service_name) as span:
```

I'm not sure how to troubleshoot from here.  Have you seen this before, or have any suggestions on what I could look at?  It's the only thread in process so nothing should be holding the GIL.

### PR DESCRIPTION
**Pull request by jd**
The current implementation it overly complicated and needs to use a Lock to be
thread safe. This new one does not.

Fixes #935

### CODE REVIEW

File: ddtrace/sampler.py [hunk 1]
```diff
@@ -50,50 +48,47 @@ def sample(self, span):
...
+        new_by_service_samplers = self._get_new_by_service_sampler()
         for key, sample_rate in iteritems(rate_by_service):
-            self._set_sample_rate_by_key(sample_rate, key)
-        with self._lock:
-            for key in list(self._by_service_samplers):
-                if key not in rate_by_service and key != _default_key:
-                    del self._by_service_samplers[key]
+            new_by_service_samplers[key] = RateSampler(sample_rate)
+
+        self._by_service_samplers = new_by_service_samplers
```
Comment (brettlangdon) [inline ddtrace/sampler.py]: hmm, this is an interesting change, I am not sure we can do this, we might need to keep any existing service/env keys around.

I am not 100% sure the agent returns all previously known service/env keys... we'll need to check the implementation in the agent or ask @gbbr

File: ddtrace/sampler.py [hunk 2]
```diff
@@ -50,50 +48,47 @@ def sample(self, span):
...
+
+
+# Default key for service with no specific rate
+RateByServiceSampler._default_key = RateByServiceSampler._key()
```
[the remark this answers is not in the document]
Comment (jd) [reply ddtrace/sampler.py]: This should be a class-wide attribute, so it _should_ not be in `__init__`.
I agree this is not optimal, but 🤷‍♂ 
I'll go with the change on L71 if you want, I think it's good enough. It's not like this function is costly.

File: ddtrace/sampler.py (hunk 1 above)
Comment (jd) [reply ddtrace/sampler.py]: If I read the previous function correctly it:
- adds set the rates for every service returned by the agent
- deletes all the rates for every service not returned by the agent

So in the end, you have only the service returned by the agent in the `_by_service_samplers`. What do I miss?

**Comment by jd** [comment]
Ok, let's go then! I think it's worth merging as it is if it fixes the issue 🤞

**Comment by apenney** [comment]
We didn't get a chance to test this, but I look forward to slipping it back in when the next release is out, I really appreciate you tackling removing this lock :)

**Pull request merged**